processing/src/main/java/io/druid/query/QueryMetrics.java

```diff
@@ -0,0 +1,198 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query;
+
+import com.metamx.emitter.service.ServiceEmitter;
+
+import java.util.Map;
+
+/**
+ * Abstraction wrapping {@link com.metamx.emitter.service.ServiceMetricEvent.Builder} and allowing to control what
+ * metrics are actually emitted, what dimensions do they have, etc.
+ *
+ *
+ * Goals of QueryMetrics
+ * ---------------------
+ *  1. Skipping or partial filtering of particular dimensions or metrics entirely. Implementation could leave the body
+ *  of the corresponding method empty, or implement random filtering like:
+ *  public void cpuTime(long timeNs)
+ *  {
+ *    if (ThreadLocalRandom.current().nextDouble() < 0.1) {
+ *      super.cpuTime(timeNs);
+ *    }
+ *  }
+ *
+ *  2. Ability to add new dimensions and metrics, possibly expensive to compute, or expensive to process (long string
+ *  values, high cardinality, etc.) and not to affect existing Druid installations, by skipping (see 1.) those
+ *  dimensions and metrics entirely in the default QueryMetrics implementations. Users who need those expensive
+ *  dimensions and metrics, could explicitly emit them in their own QueryMetrics.
+ *
+ *  3. Control over the time unit, in which time metrics are emitted. By default (see {@link DefaultQueryMetrics} and
+ *  it's subclasses) it's milliseconds, but if queries are fast, it could be not precise enough.
+ *
+ *  4. Control over the dimension and metric names.
+ *
+ *
+ * Types of methods in this interface
+ * ----------------------------------
+ *  1. Methods, pulling some dimensions from the query object. These methods are used to populate the metric before the
+ *  query is run. These methods accept a single `QueryType query` parameter. {@link #query(Query)} calls all methods
+ *  of this type, hence pulling all available information from the query object as dimensions.
+ *  2. Methods for setting dimensions, which become known in the process of the query execution or after the query is
+ *  completed.
+ *  3. Methods to register metrics to be emitted later in bulk via {@link #emit(ServiceEmitter)}. These methods
+ *  return this QueryMetrics object back for chaining.
+ *
+ *
+ * Implementors expectations
+ * -------------------------
+ * QueryMetrics is expected to be changed often, in every Druid release. Users who create their custom implementations
+ * of QueryMetrics should be ready to fix the code of their QueryMetrics (implement new methods) when they update Druid.
+ * Broken builds of custom extensions, containing custom QueryMetrics is the way to notify users that Druid core "wants"
+ * to emit new dimension or metric, and the user handles them manually: if the new dimension or metric is useful and not
+ * very expensive to process and store then emit, skip (see above Goals, 1.) otherwise.
+ *
+ * <p>QueryMetrics is designed for use from a single thread, implementations shouldn't care about thread-safety.
```

Comment:
`QueryMetrics` objects exist within query execution by design. All queries performed in a single thread, except GroupBy AFAIK, it is supposed that if GroupBy want to report some metrics from the code which is executed in parallel, it should transmit this information to the primary thread which was assigned to execute this GroupBy query, gather it and report via `QueryMetrics` object which is used only within this thread. I. e. it should be the headache of  GroupBy query implementors, not `QueryMetrics` implementors. `GroupByQueryMetrics` methods should be designed accordingly.

I don't understand how it is related to the nature of emitter